net/mqtt: Enable TLS support

askawu opened the pull request:
1. Convert MQTT lib to use net app API and modify MQTT related tests
2. Enable TLS support of MQTT lib

TODO:
1. Handle PSK and CA for mqtt_publisher
2. Add test cases for MQTT with TLS

Commit:
net/mqtt: Convert mqtt lib to use net app API

Use net app API since we want to enable MQTT with TLS.
mqtt_connect() and mqtt_close() are added to build and close the
connection to the broker. The caller doesn't need to deal with
the net context anymore and the most of network setup code in
mqtt_publisher is removed.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

Commit:
net/mqtt: Enable TLS support

CONFIG_MQTT_LIB_TLS is introduced to enable TLS support.

Also, prj_frdm_k64f_tls.conf is added to demostrate the whole idea.

jira:ZEP-2261

Signed-off-by: Aska Wu <aska.wu@linaro.org>

jukkar commented on samples/net/mqtt_publisher/prj_frdm_k64f.conf:
```diff
@@ -25,6 +25,7 @@ CONFIG_NET_IPV6=y
 CONFIG_MQTT_LIB=y
 
 CONFIG_NET_APP_SETTINGS=y
+CONFIG_NET_APP_CLIENT=y
```
This is not needed as you have that automatically selected when CONFIG_MQTT_LIB is enabled.

askawu commented:
@pfalcon @jukkar please help to review my patch

jukkar commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -420,57 +393,32 @@ static int network_setup(struct net_context **net_ctx, const char *local_addr,
-	if (rc) {
-		printk("net_context_connect error\n");
-		goto lb_exit;
+	p = net_if_ipv4_addr_add(net_if_get_default(),
```
If you call net_app_init(), then that function will make sure that the IP addresses are set to the network interface and you can remove this and the corresponding ipv6 call from the app.

jukkar commented on tests/net/lib/mqtt_publisher/src/test_mqtt_publish.c:
```diff
@@ -327,96 +326,42 @@ static int test_disconnect(void)
-	if (rc) {
-		TC_PRINT("set_addr (server) error\n");
+#else
+	rc = net_addr_pton(AF_INET6, CONFIG_NET_APP_MY_IPV4_ADDR, &addr);
```
There should be no need to set the IP addresses manually, just call net_app_init() and it will handle all of this setup.

jukkar commented on tests/net/lib/mqtt_subscriber/src/test_mqtt_subscribe.c:
```diff
@@ -364,97 +369,43 @@ static int test_disconnect(void)
-	if (rc) {
-		printk("net_context_connect error\n"
-				"Is the server (broker) up and running?\n");
+	p = net_if_ipv4_addr_add(net_if_get_default(),
```
same comment to this too, use net_app_init() here

jukkar commented on ext/lib/crypto/mbedtls/configs/config-mini-tls1_2.h:
```diff
@@ -74,7 +74,7 @@
 #endif
 
-#define MBEDTLS_SSL_MAX_CONTENT_LEN  1500
+#define MBEDTLS_SSL_MAX_CONTENT_LEN  2500
```
Please put this into separate patch. Even better would be to add if defined(MQTT...) here if this increase is required by MQTT only.

jukkar commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -61,15 +61,41 @@ struct mqtt_client_ctx {
+#define TLS_REQUEST_BUF_SIZE 1500
+#define TLS_PRIVATE_DATA "Zephyr TLS mqtt publisher"
+
+#if !defined(CONFIG_NET_APP_TLS_STACK_SIZE)
```
The TLS stack size value is set by default so no need to set it here too. If the default stack size is too long, then its value can be tweaked in prj conf file. I noticed that I had that setup like this in echo-client too but I will need to change that.

jukkar commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -61,15 +61,41 @@ struct mqtt_client_ctx {
+
+int setup_cert(struct net_app_ctx *ctx, void *cert)
+{
+	/* TODO: PSK and CA verification */
```
Before TLS support patch can be accepted, this part of the code needs to be implemented too.

jukkar commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -243,45 +269,38 @@ static void publisher(void)
+	client_ctx.mqtt_ctx.cert_host = TLS_SNI_HOSTNAME;
+	client_ctx.mqtt_ctx.tls_mem_pool = &tls_mem_pool;
+	client_ctx.mqtt_ctx.tls_stack = tls_stack;
+	client_ctx.mqtt_ctx.tls_stack_size = CONFIG_NET_APP_TLS_STACK_SIZE;
```
You need to use K_THREAD_STACK_SIZEOF() macro here like this

...tls_stack_size = K_THREAD_STACK_SIZEOF(tls_stack);

jukkar reviewed (changes requested):
Thanks for submitting this. I had this in my todo list but fortunately you got there first.

Commit:
ext: lib: mbedtls: Increase max content len if MQTT TLS is enabled

Increase MBEDTLS_SSL_MAX_CONTENT_LEN to 2500 to avoid TLS handshake
failure with mosquitto.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

askawu commented:
New patches are submitted
1. Remove network setup code since it will be done by net_app_init()
2. Use K_THREAD_STACK_SIZEOF() for the stack size
3. Handle PSK and CA verification. Also, I found a bug of mosquitto that PSK with leading zero will result TLS handshake failure.  See https://github.com/eclipse/mosquitto/pull/498 for more info.
4. Add a new changes for mbedtls config

Commit:
tests: mqtt: Add a mqtt tls test case

This adds a "build-only" test case to check if any compile error
when CONFIG_MQTT_LIB_TLS is enabled.

Signed-off-by: Aska Wu <aska.wu@linaro.org>

askawu commented:
Update:
Add a "build-only" test case to ensure there's no build error if CONFIG_MQTT_LIB_TLS is enabled.

jukkar reviewed (commented):
Looks quite good, minor comment thou.

jukkar commented on samples/net/mqtt_publisher/src/main.c:
```diff
@@ -61,15 +61,69 @@ struct mqtt_client_ctx {
+
+/* TLS */
+#define TLS_SNI_HOSTNAME "localhost"
+#define TLS_REQUEST_BUF_SIZE 1500
```
So I was just wondering here why would we need to have 2500 byte long mbedtls setting if we always read max 1500 bytes long data (size of the request buf).

askawu replied:
Good question, I will figure it out. Thanks

askawu replied:
I think MBEDTLS_SSL_MAX_CONTENT_LEN is related to the maximum size of SSL record, According to the SSL protocol, it can be 16384 at most. During the handshake with mosquitto, the SSL record is larger than 1500, this is why handshake failed.

TLS_REQUEST_BUF_SIZE is the buffer size we used to hold the data for mbedtls_ssl_read() and mbedtls_ssl_write(). But it should be less than MBEDTLS_SSL_MAX_CONTENT_LEN due to the protocol limitation.

askawu commented:
Rebased on latest upstream/master

jukkar reviewed (approved):
LGTM
